# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
table = pd.pivot_table(data[['children','debt']].rename(columns={'children':'Кол-во детей'}), values='debt', index='Кол-во детей', aggfunc={'debt': ['sum', 'count']})
table = table.rename(columns={'sum':'Кол-во должников', 'count':'Кол-во кредитополучателей'})
table['Доля должников'] = table['Кол-во должников']/table['Кол-во кредитополучателей']
table

Кол-во кредитополучателей  Кол-во должников  Доля должников
Кол-во детей                                                             
0                                 14091              1063        0.075438
1                                  4808               444        0.092346
2                                  2052               194        0.094542
3                                   330                27        0.081818
4                                    41                 4        0.097561
5                                     9                 0        0.000000

**Вывод:** Люди, не имеющие детей, чаще не имеют  задолженностей  по кредиту.  Люди, у которых есть дети, чаще имеют задолженность, причем увеличение количества детей повышает вероятность наличия задолженности, но незначительно.
Вероятность, что человек без детей, не вернет кредит в срок, равна приблизительно 7,5%.
Вероятность, что человек, имеющий детей, не вернет кредит в срок, равна приблизительно 8,2-9,8%. Причем значение для кредитора с тремя детьми отличается от значений для кредиторов с другим количеством детей, что может быть связано с небольшим количеством опрошенных, которые отнесли себя  к этой категории (по сравнению с количеством опрошенных  в других категориях) и, соответсвенно, неточным результатом.
Категорию "кредитор с 5 детьми" рассматривать нецелесообразно, так как к этой категории относится всего 9 человек.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
table = pd.pivot_table(data[['family_status','debt']].rename(columns={'family_status':'Статус'}), values='debt', index='Статус', aggfunc={'debt': ['sum', 'count']})
table = table.rename(columns={'sum':'Кол-во должников', 'count':'Кол-во кредитополучателей'})
table['Доля должников'] = table['Кол-во должников']/table['Кол-во кредитополучателей']
table

Кол-во кредитополучателей  Кол-во должников  \
Статус                                                               
Не женат / не замужем                       2796               273   
в разводе                                   1189                84   
вдовец / вдова                               951                63   
гражданский брак                            4134               385   
женат / замужем                            12261               927   

                       Доля должников  
Статус                                 
Не женат / не замужем        0.097639  
в разводе                    0.070648  
вдовец / вдова               0.066246  
гражданский брак             0.093130  
женат / замужем              0.075606

**Вывод:** Люди, находящиеся или бывшие в браке, чаще возвращают кредит в срок. 

Есть гипотеза, что имеет влияние возраст кредитора, так как самой "благонадежной" категорией оказалась категория "вдовец/вдова", а самыми "неблагонадежными" категории "Не женат/не замужем", "гражданский брак".

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
table = pd.pivot_table(data[['total_income_category','debt']].rename(columns={'total_income_category':'Категория дохода'}), values='debt', index='Категория дохода', aggfunc={'debt': ['count', 'sum']})
table = table.rename(columns={'sum':'Кол-во должников', 'count':'Кол-во кредитополучателей'})
table['Доля должников'] = table['Кол-во должников']/table['Кол-во кредитополучателей']
table

Кол-во кредитополучателей  Кол-во должников  Доля должников
Категория дохода                                                             
A                                        25                 2        0.080000
B                                      5014               354        0.070602
C                                     15921              1353        0.084982
D                                       349                21        0.060172
E                                        22                 2        0.090909

**Вывод:** Версия 1.
Нет зависимости между уровнем дохода и возвратом кредита в срок.

Версия 2.
0–30000 — 'E';
30001–50000 — 'D';
50001–200000 — 'C';
200001–1000000 — 'B';
1000001 и выше — 'A'.

По категориям А, D, E крайне мало данных, чтобы делать выводы. По категориям B и C можно сказать, что при доходе 200001–1000000 количество должников меньше, чем при доходе 50001–200000, то есть в диапазоне дохода "средний - выше среднего" количество должников с увеличением дохода уменьшается.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
# версия 1
data.groupby('purpose_category')['debt'].mean()

# версия 2
table = pd.pivot_table(data[['purpose_category','debt']].rename(columns={'purpose_category':'Цель заема'}), values='debt', index='Цель заема', aggfunc={'debt': ['sum', 'count']})
table = table.rename(columns={'sum':'Кол-во должников', 'count':'Кол-во кредитополучателей'})
table['Доля должников'] = table['Кол-во должников']/table['Кол-во кредитополучателей']
table

Кол-во кредитополучателей  Кол-во должников  \
Цель заема                                                              
операции с автомобилем                         4279               400   
операции с недвижимостью                      10751               780   
получение образования                          3988               369   
проведение свадьбы                             2313               183   

                          Доля должников  
Цель заема                                
операции с автомобилем          0.093480  
операции с недвижимостью        0.072551  
получение образования           0.092528  
проведение свадьбы              0.079118

**Вывод:** Вероятность задолженности выше, если кредит берется для целей: "операции с автомобилем", "получение образования".
Ниже вероятность задолженности для кредита, взятого для цели "проведение свадьбы".
Самая низкая вероятность задолженности для цели "операции с недвижимостью".

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

1) Отсутствие информации (информация такого рода не собиралась/не везде собиралась или определенные данные можно было не заполнять).

2) Отсутствие информации у заемщика по какому-либо вопросу

3) Невозможность ввода информации определенного типа (например, числового с плавающей запятой)

4) Человеческий фактор (например, информация была случайно удалена оператором или не собрана по причине невнимательности).

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Игнорировать строки с пропусками - плохое решение, так как в строках с пропусками есть другие не пропущенные данные, которые могут быть полезны.

Заменять пропуски нулями - плохое решение, так как число 0 для количественных переменных будет влиять на статистические характеристики.

Заменять пропуски средним (арифметическим) значением - плохое решение, так как на среднее значение очень сильно влияют выбросы  (значения, сильно отличающееся от значений выборки) и, следовательно, среднее значение может очень сильно отличаться от большинства значений в выборке и влиять на исследуемые статистические характеристики.

Заменять пропуски медианным значением - лучшее решение для количественных переменных, так как таким образом строки с пропусками не исключаются (и  вся полезная информация будет учтена) и заполненные медианным значением пропуски не будут влиять на статистические характеристики, так как будут являться усредненным большинства значений.

### Шаг 4: общий вывод.

**Общий вывод**

*Версия 1*

Самыми "благонадежными" (чаще выплачивающими кредит в срок) являются люди без детей, находящиеся или бывшие в браке,  которые получают кредит для операций с недвижимостью. Наиболее важно наличие или отсутсвие детей, количество детей влияет на  "благонадежность" слабо. 

Уровень дохода не влияет на наличие/отсутсвие задолженностей по кредиту.

Самыми "неблагонадежными" оказались холостые люди, которые берут кредит для операций с автомобилем и на образование.


*Версия 2*

В данных были найдены пропуски: 
 - 2174 пропущенных значения в days_employed (общий трудовой стаж в днях);
 - 2174 пропущенных значения в total_income (ежемесячный доход).
 
Пропуски были заменены медианными значениями по каждому типу из столбца income_type (тип занятости: сотрудник, пенсионер, студент и т.д.).

В days_employed отрицательные значения были заменены на положительные, равные по модулю исходным.

Были добавлены столбцы категорий:
 - дохода total_income_category:

Доход          | Категория
:------------- | :---------------- 
0–30000        | E      
30001–50000    | D
50001–200000   | C
200001–1000000 | B
1000001 и выше | A

* цели кредита purpose_category:
    * операции с автомобилем,
    * операции с недвижимостью,
    * проведение свадьбы,
    * получение образования.

-------------------------------------------------------------

Удалось обнаружить следующие зависимости:
 - у заемщиков с детьми больше вероятность просрочки 

Кол-во детей  | Вер-сть просрочки
:------------ | :---------------- 
Нет детей     | 7.5%        
1-5 детей     | 8.2-9.8%  

 - заемщики, состоящие или состоявшие в браке реже допускают просрочку
 
Статус                    | Вер-сть просрочки
:------------------------ | :---------------- 
Состоят (-ли) в браке     | 6.7-7.6%        
Не состоят (-ли) в браке  | 9.3-9.8%

 - ипотечники являются самой обширной категорией и самой добросовестной
 
  Цель кредита            | Кол-во заемщиков  | Вер-сть просрочки
:------------------------ | :---------------- | :----------------
Операции с недвижимостью  | 10751             | 7.3%
Проведения свадьбы        | 2313	          | 7.9%
Получение образования     | 3988              | 9.3%
Операции с автомобилем    | 4279              | 9.3%

 - при доходе 200 тыс. - 1 млн. вероятность просрочки ниже, чем при доходе 50 - 200 тыс. (по остальным категориям дохода слишком мало данных, чтобы делать выводы)
 
Доход             | Вер-сть просрочки
:---------------- | :---------------- 
50 - 200 тыс.     | 8.5%        
200 тыс. - 1 млн. | 7%